### ISSUE TITLE
Issue with v9.61 in counterwallet+counterblock

### ISSUE THREAD

**Issue by jotapea** [CounterpartyXCP/counterparty-lib#1294]
*From @jdogresorg in a chat:*

"""
For the past week or so counterwallet has been down because counterblock (counterwallets backend) has been choking on parsing blocks.

There is a problem introduced with the dispensers close delay. Everything works correctly in counterparty-lib except for the block_index in the message bindings.  Instead of inserting the block_index of the actual closure, the block_index of the original tx that marked the dispenser to be close is used. This is generating a non-correlated block_index in the messages table, so now it cannot be assumed that greater the messsage_index equal or greater is the block_index.

Javier and I spent the past week looking into the counterblock parsing issue and trying out potential fixes to get counterblock parsing again without the need for an update to counterparty-lib. However, we have come to the conclusion that in order to get counterblock parsing again, the simplest solutions are to make updates to counterparty-lib.

Neither Javier or I feel this issue rises to the level required for us to do an immediate hotfix release without community/dev feedback (ie. not an emergency).

We have put together 3 PRs and present 2 potential solutions.

Solution #1
- Update counterparty-lib to correct block_index correlation (messages ledger will change)
    https://github.com/CounterpartyXCP/counterparty-lib/pull/1293

Solution #2
- Update counterparty-lib get_blocks function to work with non-correlated block_index
    https://github.com/CounterpartyXCP/counterparty-lib/pull/1292

- Update counterblock to handle the dispenser close delay.
    https://github.com/CounterpartyXCP/counterblock/pull/196

It is up to the community and the core maintainers how they would like to address this issue.
"""

*Just taking the liberty of making it into a protocol code repository issue, like I believe these should be done.*

**Comment by jotapea** [comment]
To start, I just want to say that the intention to be more transparent is greatly appreciated.

Second, I want to be clear in learning who you are referring to with the "core maintainers" (who aside from @jdogresorg and @pataegrillo).

Then about the issue itself, I'm a bit confused by the wording used. Why would the protocol be changed for an issue in a client application? It is either an issue at the protocol, or at the application...

**Comment by jotapea** [comment]
Ok after re-reading and looking at the code it is clear the issue is in the protocol. IMO you should just issue another hotfix (v9.61.2) ASAP and avoid having to add another protocol change. Or disable the delay dispenser close protocol change.

We need to be consistent in the protocol version numbers. Meaning that adding this protocol change will mean v9.62.

Hopefully [these](https://github.com/CounterpartyXCP/counterparty-lib/pull/1284) are lessons for not rushing a [release](https://github.com/CounterpartyXCP/cips/discussions/124) unnecessarily.

**Comment by jdogresorg** [comment]
The current "issue" with non-correlated `block_index` in messages currently only effects `counterblock` (a 10 year old outdated / external component). 

`counterparty-lib` 9.61.1 is working perfectly fine, all transactions and ledger interactions are executing as expected.

Automated testing, which I have been screaming that CP needs for years, would have most likely caught this issue in counterblock parsing. However, since all testing is manual, all that I have been testing is the individual component being changed (ie, if I am changing `counterparty-lib`, I test `counterparty-lib` to make sure it functions as expected and stores data internally as it should)

Javier and I have looked into this "issue" for over a week (at no cost to the community) and provided 2 possible solutions (opinions AND code) to the community to get `counterblock` working again if they feel that is a priority.

All details of the issue, the solutions, and code to back up those solutions is documented on github in PRs.. That is all I have to say about this "issue". 

The community and `core` maintainers can decide what is best for the project and push it forward on their own.

On a personal note: 

NO release was rushed... Stop pushing this false narrative... You can think what you want, but the fact is, the same processes have been followed for the past 5+ years were used in this release... regardless of how many times you repeat this false narrative, or link to lots of your own github issues/PRs/Posts where you try and push this false narrative,  it does not make it true.

Perhaps focus on moving CP dev forward and less of the snarky "I told you so" type nonsense that has caused me to step away from Counterparty Development and focus my dev efforts elsewhere. I am glad you have a bunch of opinions... but perhaps you should try contributing code and putting forth your own ideas and pushing them forward rather than focusing all your time being passive-agressive and attacking others who are doing the building.

TLDR... easy to criticize... hard to build... Glad you got it all figured out and know blockchain/dev better than those of us 
who have been in software development 30+ years and blockchain dev 10+ years... looking forward to seeing how you push CP dev forward now that i am "out of the way" 🤣

![candy](https://github.com/CounterpartyXCP/counterparty-lib/assets/376028/d053cb73-e6a2-4b5b-9875-ae967f25c689)

**Comment by jotapea** [comment]
Wow perfect [timing](https://github.com/CounterpartyXCP/counterparty-lib/issues/1227#issuecomment-1858857660) I guess?

Anyway, I don't agree with your starting point:

> The current "issue" with non-correlated block_index in messages currently only effects counterblock (a 10 year old outdated / external component).

Real counterparty application developers (node runners, not api consumers) rely on the `messages` table. It is THE MAIN table used if you want to build your own application-level DB, separate from the protocol consensus validity SQLite.

I'm glad being someone bringing in a new perspective. It has been way more difficult than it needs to, but I understand that is not easy to let go of full control. I'm here to [Decentralize CNTRPRTY](https://github.com/CNTRPRTY/), and each day more people appreciate this.

@pataegrillo (and any other of the "core maintainers", which I'm still not sure if there are more...), do you think this should be merged asap to become v9.61.2, be left to v9.62, or what else? I believe the MAIN issue here is HOW to handle this, not the bug itself...

**Comment by jdogresorg** [comment]
> Wow perfect timing I guess?

Dunno what your referring to, nor do I care to listen anymore.... You got it all figured out... good luck. 👍 

> @pataegrillo (and any other of the "core maintainers", which I'm still not sure if there are more...)

Yes... your unaware of a bunch of things going on with counterparty, and have been for quite a while... because you refuse to  engage with the other CP developers in real-time exchanges except in your echo-chamber chat rooms. When you have been in the official CP and dev channels with hundreds of devs,  multiple people have tried to communicate to you that your development / communication style could use some improvements. You don't like that feedback so you quit those channels and setup your own channels under the guise of "decentralizing" counterparty, and continue your passive-agressive attacks on github, and then always act confused here on github like your left out of conversations....

IMO your true goal is to divide the community and concern troll.

Core maintainers with commit access are @jdogresorg @Jpja @genecyber @wasthatawolf @robby-dermody .

![image](https://github.com/CounterpartyXCP/counterparty-lib/assets/376028/1f1881c7-9bed-439b-bf02-0cef3a0d21f8)

![image](https://github.com/CounterpartyXCP/counterparty-lib/assets/376028/db92be15-69e5-4612-83db-c5df69809e32)

**Comment by jotapea** [comment]
We are more aligned than you think, but we do differ a lot in approach and perspective.

Thanks for the @ and images, very helpful.

**Comment by robby-dermody** [comment]
IMO this goes beyond counterblock. anything consuming counterparty-lib messages should not have to make a separate assertion (and separate logic) for dispensers. I vote for Solution #1.

**Comment by adamkrellenstein** [comment]
100% agreed with @robby-dermody. This is a protocol-level bug and it needs to be addressed accordingly.

**Comment by adamkrellenstein** [comment]
Okay, seems like the best solution is to fix the ledger (esp. given the fact that credits/debits won't change! though we'll have to test that ofc.) Going with #1.
